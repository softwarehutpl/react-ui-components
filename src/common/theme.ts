const theme: { [key: string]: { [key: string]: { [key: string]: string } } } = {
  colors: {
    primary: {
      base: '#000000',
      light: '#ffffff',
      dark: '#a8b3bd',
    },
    secondary: {
      base: '#3690e3',
      light: '#75b0e6',
      dark: '#034d91',
    },
    error: {
      base: '#f44336',
      light: '#f56c62',
      dark: '#a1140a',
    },
    warning: {
      base: '#faa72d',
      light: '#ffc36b',
      dark: '#995b00',
    },
    info: {
      base: '#7979d4',
      light: '#9797d1',
      dark: '#4c4cc7',
    },
    success: {
      base: '#4caf50',
      light: '#88c28a',
      dark: '#124f14',
    },
    disabled: {
      base: '#d9d9d9',
      light: '#f0eded',
      dark: '#8c8989',
    },
  },
  // TODO: change it in dedicated task for that (change ease-out for example and add more options)
  transitions: {
    all: {
      fast: 'all 0.5s ease-out',
      mid: 'all 2s ease-out',
      slow: 'all 4s ease-out',
    },
    background: {
      fast: 'background-color 0.5s ease-out',
      mid: 'background-color 2s ease-out',
      slow: 'background-color 4s ease-out',
    },
    fontColor: {
      fast: 'color 0.5s ease-out',
      mid: 'color 2s ease-out',
      slow: 'color 4s ease-out',
    }
  }
};

export default theme;
