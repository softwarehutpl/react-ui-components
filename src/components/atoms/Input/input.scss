@import 'src/common/styles/variables.scss';

.test-class {
  color: $color-black;
  background-color: $color-white;
  border-color: $color-cerise;
}