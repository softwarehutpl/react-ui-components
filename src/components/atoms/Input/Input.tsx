import React from 'react';
import styled from 'styled-components';
import './input.scss';

type Props = {
  buttonTitle?: string;
  onChange: () => void;
  color?: string;
  className?: string;
  placeholder?: string;
  disabled?: boolean;
  fontColor?: string;
  noBorder?: boolean;
  borderColor?: string;
  hoverBorderColor?: string;
  focusBorderColor?: string;
  borderWidth?: number;
  borderRadius?: number;
  backgroundColor?: string;
  hoverBackgroundColor?: string;
  focusBackgroundColor?: string;
  transitionEffect?: string;
  showPlaceholderOnFocus?: boolean;
  margin?: number;
  padding?: number;
}

const Input = (
  {
    onChange,
    className,
    placeholder,
    disabled,
  }: Props
) => (
  <input
    className={className}
    onChange={onChange}
    disabled={disabled}
    placeholder={placeholder}
  />
);

const defaultProps = {
  color: 'primary',
  fontColor: '#000',
  noBorder: false,
  disabled: false,
  className: '',
  borderWidth: 2,
  margin: 0,
  padding: 15,
  borderRadius: 0,
  borderColor: '#000',
  showPlaceholderOnFocus: false,
};

Input.defaultProps = defaultProps;

export default styled(Input)`${({ 
  theme,
  color = defaultProps.color,
  backgroundColor,
  fontColor = defaultProps.fontColor,
  noBorder = defaultProps.noBorder,
  borderColor = defaultProps.borderColor,
  borderWidth = defaultProps.borderWidth,
  borderRadius = defaultProps.borderRadius,
  margin = defaultProps.margin,
  padding = defaultProps.padding,
  hoverBackgroundColor,
  hoverBorderColor,
  focusBackgroundColor,
  focusBorderColor,
  transitionEffect,
  showPlaceholderOnFocus,
}) => ({
  'background-color': backgroundColor || theme.colors[color].light,
  'color': fontColor || theme.colors[color].dark,
  'border': noBorder ? 'none' : 'solid',
  'border-color': borderColor || theme.colors[color].dark,
  'border-width': `${borderWidth}px`,
  'border-radius': `${borderRadius}px`,
  'margin': `${margin}px`,
  'padding': `${padding}px`,
  '&:hover': {
    'cursor': 'pointer',
    'background-color': hoverBackgroundColor || focusBackgroundColor || theme.colors[color].light,
    'border-color': hoverBorderColor || focusBorderColor || theme.colors[color].dark,
  },
  '&:focus': {
    'cursor': 'pointer',
    'background-color': focusBackgroundColor || theme.colors[color].light,
    'border-color': focusBorderColor || theme.colors[color].dark,
  },
  '&:focus::-webkit-input-placeholder': showPlaceholderOnFocus ? {
    'font-size': '.75em',
    'position': 'relative',
    'top': '-15px',
    'transition': transitionEffect ? theme.transitions.all[transitionEffect] : '0.2s ease-out',
  } : '',
  '&::-webkit-input-placeholder': showPlaceholderOnFocus ? {
    'transition': transitionEffect ? theme.transitions.all[transitionEffect] : '0.2s ease-out',
  } : '',
  'outline': 'none',
  'transition': transitionEffect ? theme.transitions.all[transitionEffect] : 'none',
  '&[disabled]': {
    'background-color': theme.colors.disabled.base,
    'border-color': theme.colors.disabled.dark,
    'color': '#a6a6a6',
    '&:hover': {
      'cursor': 'auto',
    },
  },
})}`
