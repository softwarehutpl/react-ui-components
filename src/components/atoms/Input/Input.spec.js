
import React from 'react';
import { create, act } from 'react-test-renderer';
import 'jest-styled-components';
import Input from './Input';
import theme from '../../../common/theme';

describe('Input Component', () => {
  it('should render correctly', () => {
    const tree = create(<Input theme={theme} placeholder='test' onChanege={() => {}} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render input component with proper placeholder', () => {
    const component = create(<Input placeholder='test' theme={theme} />);
    const instance = component.root;
    const input = instance.findByType('input');
    expect(input.props.placeholder).toBe('test');
  });

  it('should call input onChange method', () => {
    const mockOnChange = jest.fn();
    const component = create(<Input onChange={mockOnChange} theme={theme} />);
    const instance = component.root;
    const input = instance.findByType(Input);
    act(() => {
      input.props.onChange();
    });
    expect(mockOnChange.mock.calls.length).toEqual(1);
  });

  it('should render input with custom class', () => {
    const component = create(<Input theme={theme} className='test-class' />);
    const input = component.toJSON();
    expect(input.props.className).toEqual(expect.stringContaining('test-class'));
  });

  it('should render input component with selected theme styling', () => {
    const component = create(<Input theme={theme} color="secondary" />);
    const input = component.toJSON();
    expect(input).toHaveStyleRule('background-color', '#75b0e6');
    expect(input).toHaveStyleRule('color', '#000');
    expect(input).toHaveStyleRule('border-color', '#000');
    expect(input).toHaveStyleRule('background-color', '#75b0e6', { modifier: ':hover' });
    expect(input).toHaveStyleRule('background-color', '#75b0e6', { modifier: ':focus' });
  });

  it('should render disabled input', () => {
    const component = create(<Input theme={theme} disabled />);
    const input = component.toJSON();
    expect(input.props.disabled).toBe(true);
  });

  it('should render input component with color based on fontColor prop', () => {
    const component = create(<Input theme={theme} fontColor='hotpink' />);
    const input = component.toJSON();
    expect(input).toHaveStyleRule('color', 'hotpink');
  });

  it('should render input with custom hover and focus styles', () => {
    const component = create(<Input
      theme={theme}
      hoverBackgroundColor='#CECECE'
      focusBackgroundColor='#C7C7C7C'
      hoverBorderColor='#454545'
      focusBorderColor='#474747'
    />)
    const input = component.toJSON();
    expect(input).toHaveStyleRule('border-color', '#454545', { modifier: ':hover' });
    expect(input).toHaveStyleRule('background-color', '#CECECE', { modifier: ':hover' });
    expect(input).toHaveStyleRule('border-color', '#474747', { modifier: ':focus' });
    expect(input).toHaveStyleRule('background-color', '#C7C7C7C', { modifier: ':focus' });
  });

  it('should render input with custom hover and focus, when no focus parameters introduced, but focus are introduced',
    () => {
    const component = create(<Input
      theme={theme}
      focusBackgroundColor='#C7C7C7C'
      focusBorderColor='#474747'
    />)
    const input = component.toJSON();
    expect(input).toHaveStyleRule('border-color', '#474747', { modifier: ':hover' });
    expect(input).toHaveStyleRule('background-color', '#C7C7C7C', { modifier: ':hover' });
    expect(input).toHaveStyleRule('border-color', '#474747', { modifier: ':focus' });
    expect(input).toHaveStyleRule('background-color', '#C7C7C7C', { modifier: ':focus' });
  });

it('should render input with custom border', () => {
    const component = create(
      <Input
        theme={theme}
        noBorder={false}
        borderColor='#CDE769'
        borderWidth={3}
        borderRadius={10}
      />
    );
    const input = component.toJSON();
    expect(input).toHaveStyleRule('border', 'solid');
    expect(input).toHaveStyleRule('border-color', '#CDE769');
    expect(input).toHaveStyleRule('border-width', '3px');
    expect(input).toHaveStyleRule('border-radius', '10px');
  });

  it('should render input component with custom transition', () => {
    const component = create(<Input theme={theme} />);
    const input = component.toJSON();
    expect(input).toHaveStyleRule('transition', 'none');
  });

  it('should render input component with transitionEffect', () => {
    const component = create(<Input theme={theme} transitionEffect='mid' />);
    const input = component.toJSON();
    expect(input).toHaveStyleRule('transition', 'all 2s ease-out');
  });

  it('should render input component with custom padding and margin', () => {
    const component = create(<Input theme={theme} padding={10} margin={20} />);
    const input = component.toJSON();
    expect(input).toHaveStyleRule('padding', '10px');
    expect(input).toHaveStyleRule('margin', '20px');
  });
});
