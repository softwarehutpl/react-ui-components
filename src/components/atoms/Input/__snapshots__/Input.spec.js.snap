// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component should render correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  width: undefinedpx;
}

<div
  className="c0"
  onChanege={[Function]}
  placeholder="test"
/>
`;
